app/controllers/sessions_controller.rb

```diff
@@ -26,4 +26,22 @@ def destroy
     log_out if logged_in?
     redirect_to root_url
   end
+
+  def omniauth
+    user = User.find_or_create_by(uid: request.env['omniauth.auth'][:uid],
+                                  provider: request.env['omniauth.auth'][:provider]) do |user|
+      user.name = request.env['omniauth.auth'][:info][:name]
+      user.email = request.env['omniauth.auth'][:info][:email]
+      user.name = request.env['omniauth.auth'][:info][:nickname] if user.name.nil?
+      user.email = user.name + '@gmail.com' if user.email.nil?
```

Comment:
Using `if user.email.nil?` is relatively risky.
Better to use `case ... when ...` with `provider`

Reply:
em đã sửa